### PR TITLE
Print slack histogram at end of placement and at end of routing

### PR DESCRIPTION
**Pull request by eddiehung**
Refactor timing.cc into a class, easing ability to collect data for slack histogram, for issue #28.

On picorv32 with `--freq 70`, this is what it now prints out the following after placement:
```
Info: estimated Fmax = 54.15 MHz

Info: Slack histogram:
Info:  legend: * represents 56 endpoint(s)
Info:  -4183 < ps <  -3316 |*
Info:  -3316 < ps <  -2449 |
Info:  -2449 < ps <  -1582 |
Info:  -1582 < ps <   -715 |**
Info:   -715 < ps <    152 |****
Info:    152 < ps <   1019 |*****
Info:   1019 < ps <   1886 |**********
Info:   1886 < ps <   2753 |****************
Info:   2753 < ps <   3620 |*****************
Info:   3620 < ps <   4487 |********************
Info:   4487 < ps <   5354 |*************************
Info:   5354 < ps <   6221 |*******************************************
Info:   6221 < ps <   7088 |************************************************************
Info:   7088 < ps <   7955 |***********************************************************
Info:   7955 < ps <   8822 |***************************************************
Info:   8822 < ps <   9689 |************************************************
Info:   9689 < ps <  10556 |****************************************
Info:  10556 < ps <  11423 |*********************
Info:  11423 < ps <  12290 |**************
Info:  12290 < ps <  13157 |***
Info:  13157 < ps <  14024 |
```

and this after routing
```
Info: estimated Fmax = 55.88 MHz

Info: Slack histogram:
Info:  legend: * represents 57 endpoint(s)
Info:  -3609 < ps <  -2771 |
Info:  -2771 < ps <  -1933 |
Info:  -1933 < ps <  -1095 |*
Info:  -1095 < ps <   -257 |**
Info:   -257 < ps <    581 |***
Info:    581 < ps <   1419 |******
Info:   1419 < ps <   2257 |********
Info:   2257 < ps <   3095 |*************
Info:   3095 < ps <   3933 |*******************
Info:   3933 < ps <   4771 |*************************
Info:   4771 < ps <   5609 |***************************************
Info:   5609 < ps <   6447 |************************************************************
Info:   6447 < ps <   7285 |**********************************************************
Info:   7285 < ps <   8123 |******************************************************
Info:   8123 < ps <   8961 |***************************************************
Info:   8961 < ps <   9799 |*******************************************
Info:   9799 < ps <  10637 |************************
Info:  10637 < ps <  11475 |**************
Info:  11475 < ps <  12313 |**********
Info:  12313 < ps <  13151 |*
Info:  13151 < ps <  13989 |
```

### CODE REVIEW

File: common/timing.cc [hunk 1]
```diff
@@ -211,9 +232,32 @@ delay_t timing_analysis(Context *ctx, bool print_fmax, bool print_path)
...
+            bin += i.second;
+            max_freq = std::max(max_freq, bin);
+        }
+        bar_width = std::min(bar_width, max_freq);
+
+        log_break();
+        log_info("Slack histogram:\n");
+        log_info(" legend: * represents %d endpoint(s)\n", max_freq / bar_width);
+        for (unsigned i = 0; i < bins.size(); ++i)
+            log_info("%6d < ps < %6d |%s\n", min_slack + bin_size * i, min_slack + bin_size * (i + 1),
```
Comment (daveshah1) [inline common/timing.cc]: As I understand it, delay_t is not guaranteed to be an integer (indeed I don't think it is in generic) . I may have previously made the same mistaken assumption in the original code, so not really your fault.

Comment (eddiehung) [reply common/timing.cc]: Good thinking batman! I'll resolve it tomorrow.

Comment (daveshah1) [reply common/timing.cc]: Thanks!

Comment (eddiehung) [reply common/timing.cc]: https://github.com/YosysHQ/nextpnr/pull/31/commits/8974ef332737bcf2f37cdd59b2c9767fa187fdf3 changes `DelayFrequency`'s key to be an `int` type, and uses `int(Arch::getDelayNS() * 1000)` for picosecond granularity.

That should do it, right?

Comment (daveshah1) [reply common/timing.cc]: Yeah, that looks good

**Pull request merged**